### PR TITLE
[Azure SignalR service]Support Private Link and add new api-version

### PR DESCRIPTION
**Pull request by juniwang**
### Changes
- The first commit copied everything from existing api-version `2018-10-01` without any change, for easier code review.  Please review later commits for real changes.
- Adding Paths, Definitions for `PrivateLink`, and several new properties of `SignalRResource`.
- Adding new api-version `2020-05-10`.
- Tools documented in https://github.com/Azure/adx-documentation-pr/wiki/Swagger-Validation-Tools are all passed.

And there was another related [PR #8540](https://github.com/Azure/azure-rest-api-specs/pull/8540), which was reviewed a few times.  On top of #8540, we added new `api-version` based on the comments. 

### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/adx-documentation-pr/wiki/Overall-basic-flow) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [x] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [x] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [x] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

### CODE REVIEW

**Comment by ChenTanyi** [comment]
@juniwang It seems the example and json is not referred. Could you please solve it?

**Comment by juniwang** [comment]
> @juniwang It seems the example and json is not referred. Could you please solve it?

Intentionally not referenced. Because we are adding a new API-version. Just copy all files from existing api-version without any change.  Next PR, we will send the real changes including correct references. In this way, it's easier to do code-review.  Otherwise, there are too many files to review

**Comment by juniwang** [comment]
Of course, we can append the real changes in this PR, if you'd like

**Comment by ChenTanyi** [comment]
If the API is not ready, it can't be merged. You can append change in this PR until it ready. Thanks!

**Comment by juniwang** [comment]
> If the API is not ready, it can't be merged. You can append change in this PR until it ready. Thanks!

appended

**Comment by juniwang** [comment]
For the LintDiff failure,  can somebody take a look? I don't get the point. What **exactly** is wrong in our paths?  Which word(s)?

```
Resource type naming must follow camel case. Path: '/subscriptions/{subscriptionId}/providers/Microsoft.SignalRService/signalR'
  at specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json:18

R3021 - PathResrouceTypeNameCamelCase
Resource type naming must follow camel case. Path: '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.SignalRService/signalR'
  at specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json:18

R3021 - PathResrouceTypeNameCamelCase
Resource type naming must follow camel case. Path: '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.SignalRService/signalR/{resourceName}/privateEndpointConnections/{privateEndpointConnectionName}'
  at specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json:18

R3021 - PathResrouceTypeNameCamelCase
Resource type naming must follow camel case. Path: '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.SignalRService/signalR/{resourceName}/privateLinkResources'
  at specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json:18

R3021 - PathResrouceTypeNameCamelCase
Resource type naming must follow camel case. Path: '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.SignalRService/signalR/{resourceName}/listKeys'
  at specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json:18

R3021 - PathResrouceTypeNameCamelCase
Resource type naming must follow camel case. Path: '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.SignalRService/signalR/{resourceName}/regenerateKey'
  at specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json:18

```

**Comment by ChenTanyi** [comment]
> For the LintDiff failure, can somebody take a look? I don't get the point. What **exactly** is wrong in our paths? Which word(s)?
> 
> ```
> Resource type naming must follow camel case. Path: '/subscriptions/{subscriptionId}/providers/Microsoft.SignalRService/signalR'
>   at specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json:18
> 
> R3021 - PathResrouceTypeNameCamelCase
> Resource type naming must follow camel case. Path: '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.SignalRService/signalR'
>   at specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json:18
> 
> R3021 - PathResrouceTypeNameCamelCase
> Resource type naming must follow camel case. Path: '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.SignalRService/signalR/{resourceName}/privateEndpointConnections/{privateEndpointConnectionName}'
>   at specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json:18
> 
> R3021 - PathResrouceTypeNameCamelCase
> Resource type naming must follow camel case. Path: '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.SignalRService/signalR/{resourceName}/privateLinkResources'
>   at specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json:18
> 
> R3021 - PathResrouceTypeNameCamelCase
> Resource type naming must follow camel case. Path: '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.SignalRService/signalR/{resourceName}/listKeys'
>   at specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json:18
> 
> R3021 - PathResrouceTypeNameCamelCase
> Resource type naming must follow camel case. Path: '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.SignalRService/signalR/{resourceName}/regenerateKey'
>   at specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json:18
> ```

@NullMDR @akning-ms Could you take a look? Does CI wrongly recognize the `signalR`?

**Comment by NullMDR** [comment]
@jianyexi Could you please help on this issue?

**Comment by jianyexi** [comment]
> > For the LintDiff failure, can somebody take a look? I don't get the point. What **exactly** is wrong in our paths? Which word(s)?
> > ```
> > Resource type naming must follow camel case. Path: '/subscriptions/{subscriptionId}/providers/Microsoft.SignalRService/signalR'
> >   at specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json:18
> > 
> > R3021 - PathResrouceTypeNameCamelCase
> > Resource type naming must follow camel case. Path: '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.SignalRService/signalR'
> >   at specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json:18
> > 
> > R3021 - PathResrouceTypeNameCamelCase
> > Resource type naming must follow camel case. Path: '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.SignalRService/signalR/{resourceName}/privateEndpointConnections/{privateEndpointConnectionName}'
> >   at specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json:18
> > 
> > R3021 - PathResrouceTypeNameCamelCase
> > Resource type naming must follow camel case. Path: '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.SignalRService/signalR/{resourceName}/privateLinkResources'
> >   at specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json:18
> > 
> > R3021 - PathResrouceTypeNameCamelCase
> > Resource type naming must follow camel case. Path: '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.SignalRService/signalR/{resourceName}/listKeys'
> >   at specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json:18
> > 
> > R3021 - PathResrouceTypeNameCamelCase
> > Resource type naming must follow camel case. Path: '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.SignalRService/signalR/{resourceName}/regenerateKey'
> >   at specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json:18
> > ```
> 
@ChenTanyi  there is a known issue on LintDiff  and we have fixed it in the  ~ Staging validation pipeline ,  you can just focus on the failure in ~ Staging validation (rest-api-specs) (LintDiff)    , it seems that the enum name  below is duplicated:
```
"x-ms-enum": {
            "name": "ServiceKind",
            "modelAsString": true
          },
```

**Comment by juniwang** [comment]
For the `x-ms-enum` for `ServiceKind`, is it required to fix? If not, please help us move forward. Changes of this PR(including the referenced one) has been reviewing for more than 2 months, lots for our work are blocked.

It's hard for us to fix it now.  Because due to historic reasons, the `Request` and `Response` of `PUT` api are not sharing the same schema, `ServiceKind` is defined in both  `Request` and `Response`. Cannot revome any of them.

**Comment by ChenTanyi** [comment]
@juniwang Can you use a common `#/definitions` to describe these two same enum? SDK may not be possible to deal with duplicate name of enum.

**Comment by jianyexi** [comment]
> For the `x-ms-enum` for `ServiceKind`, is it required to fix? If not, please help us move forward. Changes of this PR(including the referenced one) has been reviewing for more than 2 months, lots for our work are blocked.
> 
> It's hard for us to fix it now. Because due to historic reasons, the `Request` and `Response` of `PUT` api are not sharing the same schema, `ServiceKind` is defined in both `Request` and `Response`. Cannot revome any of them.

Could you rename one of them, or merge them into  a common object ?

**Comment by juniwang** [comment]
From https://github.com/AzureSDKAutomation/azure-sdk-for-net/pull/1046, the .Net SDK was generated successfully.

Could you please give an example how to define properties of `x-ms-enum` in `#/definitions/`?

**Comment by ChenTanyi** [comment]
@juniwang I'm not sure about the `#/definitions`, but I found many in `#/parameters`. https://github.com/azure/azure-rest-api-specs/blob/84c087e5662c448b3ffe39263220ed77d18c9003/specification/apimanagement/control-plane/Microsoft.ApiManagement/preview/2017-03-01/apimtenant.json#L664 Not sure is it the same. 

Actually for java SDK, we will ignore for same enum which will not cause error. Will it any other language generate failed due to the duplicate enum? @jianyexi

**Comment by jianyexi** [comment]
> @juniwang I'm not sure about the `#/definitions`, but I found many in `#/parameters`. https://github.com/azure/azure-rest-api-specs/blob/84c087e5662c448b3ffe39263220ed77d18c9003/specification/apimanagement/control-plane/Microsoft.ApiManagement/preview/2017-03-01/apimtenant.json#L664 Not sure is it the same.
> 
> Actually for java SDK, we will ignore for same enum which will not cause error. Will it any other language generate failed due to the duplicate enum? @jianyexi

As I know,if the duplicate enum have different entries , the code gen will failed. This PR seems normal  But on the other way, if two enum have the same entries and name, we should use a common definition , e.g. https://github.com/azure/azure-rest-api-specs/blob/84c087e5662c448b3ffe39263220ed77d18c9003/specification/azsadmin/resource-manager/azurebridge/Microsoft.AzureBridge.Admin/preview/2016-01-01/AzureBridge.json#L483

**Comment by juniwang** [comment]
Thanks all for the replies. We fixed the `x-ms-enum` issue. 
But looks we have another issue:  `No readme.md specification configuration files were found that are associated with the files modified in this pull request, or swagger_to_sdk section in readme.md is not configured`.  Please help

**Comment by NullMDR** [comment]
@juniwang There's a bug related to azure pipeline check update. Please add a new empty commit or close and open a new PR to mitigate.

**Comment by juniwang** [comment]
For the DotNET SDK build failure, it's kind of as expected. The error message is `error CS0246: The type or namespace name 'SignalRUpdateParameters' could not be found`, that's because in this PR we in deed deleted this type, which was used as input for `SignalR_CreateOrUpdate`(`PUT`), meanwhile the output of this API was using another schema `SignalRResource`, the input and output schema were inconsistent. So we did some refactoring, in future, we will use the same schema `SignalRResource` for both input and output to conform ARM spec, the input schema `SignalRUpdateParameters` is no longer needed.

However our test case in .Net SDK are still using `SignalRUpdateParameters` which causes the build failure. We will fix the test cases and manually send PR to .Net SDK repo after this PR merged.

**Comment by juniwang** [comment]
Can one of the admin review this PR?

**Comment by ChenTanyi** [comment]
@juniwang Could you please fix the lint CI error first? following the https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/ci-fix.md

It seems the bot have wrongly removed the tags, I've added it again.
The api will need to wait for the ARM review. If you want to ask them for quick response, you could follow the first comment or send email to armapireview@microsoft.com.

**Comment by juniwang** [comment]
Could you please tell me which enum should be fixed? The logs 
"Must not have duplicate name of x-ms-enum extension , make sure every x-ms-enum name unique.
  at specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json:778" doesn't have concrete info. Looks like line 778 in that file is not relevent

**Comment by juniwang** [comment]
Tried to search the file, and found `ProvisioningState` might be the one reported in Lint. Does this one actually break something? because I do see similar cases in other specs: [Container Registry](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2019-05-01/containerregistry.json#L1954), [Kusto](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/azure-kusto/resource-manager/Microsoft.Kusto/stable/2020-02-15/kusto.json#L2459), [Batch AI](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/batchai/resource-manager/Microsoft.BatchAI/stable/2018-05-01/BatchAI.json#L3785) and others.

If it's not required, we won't change it. Please clarify why this matters.

**Comment by ChenTanyi** [comment]
I think the above have discussed about it. You can use a common `definition` for it.

I think it is the new CI in automation. It is better to follow it since some service will use same name to describe different enum, which will break the sdk generation.

**Comment by ChenTanyi** [comment]
@jianyexi Can you add additional information about the CI?

**Comment by jianyexi** [comment]
@juniwang    Better to use a common definition for  `provisioningState`  at  line 1177 and 1456  , or you can change the x-ms-enum name for any of them .

**Comment by juniwang** [comment]
the PR has been reviewing for 3 month， do we have a SLA for this?

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

File: specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json
Comment (pilor) [inline specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json:284]: it should just be 200 since only updates are allowed on privateEndpointConnection resources, not creates.  Async will still work with 200 + provisioningState

File: specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json
Comment (pilor) [inline specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json:294]: >Create [](start = 45, length = 6)

Update

File: specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json
Comment (pilor) [inline specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json:1310]: >"SignalRNetworkACLs [](start = 4, length = 19)

if any of the properties in here are required when specifying signalRNetworkAcls please mark them as required

File: specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json
Comment (pilor) [inline specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json:1369]: >ClientConnection, ServerConnection, RESTAPI [](start = 82, length = 43)

if the values are bounded please set items to a string enum

File: specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json
Comment (pilor) [inline specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json:1394]: is this a reference to a privateEndpointConnection?  If so shouldn't it use the full resource ID of the privateEndpointConnection?

File: specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json
Comment (pilor) [inline specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json:1396]: >ClientConnection, ServerConnection, RESTAPI. [](start = 82, length = 44)

same string enum comment as the networkacl definition

**Review by pilor — CHANGES_REQUESTED** [review]
:clock1:

File: specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json
Comment (juniwang) [reply specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json]: Updated. 201 removed

File: specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json
Comment (juniwang) [reply specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json]: Updated to `SignalRPrivateEndpointConnections_Update`

File: specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json
Comment (juniwang) [reply specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json:1397]: Updated to use `enum`

File: specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json
Comment (juniwang) [reply specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json]: Updated as well

File: specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json
Comment (juniwang) [reply specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json:1334]: Already done for required properties.

File: specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json
Comment (juniwang) [reply specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json]: `name` is also unique by definition. It has exactly the same effect as resource id.

Comment (pilor) [reply specification/signalr/resource-manager/Microsoft.SignalRService/stable/2020-05-01/signalr.json]: references should nearly always be by ID.  You don't know what additional scenarios are going to arise in the future that would make name not unique.  If you feel strongly that you want to reference by name please update the property name to something more explicit.  Right now the object reads as if this is the "name" of the PrivatEndpointAcl, not a reference by name to a private endpoint connection.

**Review by pilor — APPROVED** [review]
Approving with 1 remaining comment

**Review by ChenTanyi — APPROVED** [review]

**Comment by ChenTanyi** [comment]
@isra-fel Could you see about the .NET generation failure? Does it need to be fixed?

**Comment by isra-fel** [comment]
LGTM

**Comment by juniwang** [comment]
Thanks for everyone for the kind and patient review. We will move forward, to fix the .Net unit tests, cli extensions and so on

**Pull request merged**
